browser/net/brave_network_delegate_base.cc

```diff
@@ -71,7 +72,14 @@ int BraveNetworkDelegateBase::OnHeadersReceived(net::URLRequest* request,
   ctx->original_response_headers = original_response_headers;
   ctx->override_response_headers = override_response_headers;
   ctx->allowed_unsafe_redirect_url = allowed_unsafe_redirect_url;
-  RunNextCallback(request, nullptr, ctx);
+
+  // Return ERR_IO_PENDING and run callbacks later by posting a task.
+  // URLRequestHttpJob::awaiting_callback_ will be set to true after we
+  // return net::ERR_IO_PENDING here, callbacks need to be run later then this
+  // to set awaiting_callback_ back to false.
+  BrowserThread::PostTask(BrowserThread::IO, FROM_HERE,
+      base::Bind(&BraveNetworkDelegateBase::RunNextCallback,
+        base::Unretained(this), request, nullptr, ctx));
```

Comment:
Do you think this needs to be done for `OnBeforeURLRequest` and `OnBeforeStartTransaction`?

Reply:
awaiting_callback_ is for onHeadersReceived only, so I wouldn't presume these two will need to do the same thing for this reason.
Since they seem to works fine in our other cases, I don't see a reason to change them now.